File: [0235][Easy][Lowest Common Ancestor of a Binary Search Tree]/Lowest_Common_Ancestor_of_a_Binary_Search_Tree.java
//Definition for a binary tree node.
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
} 
// solution 1, use recursion to travel all the nodes of the BST.
// leetcode time     cost : 7 ms
// leetcode memory   cost : 37.5 MB
// Time  Complexity: O(N)
// Space Complexity: O(N)
class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {

        // Value of current node or parent node.
        int parentVal = root.val;

        // Value of p
        int pVal = p.val;

        // Value of q;
        int qVal = q.val;

        if (pVal > parentVal && qVal > parentVal) {
            // If both p and q are greater than parent
            return lowestCommonAncestor(root.right, p, q);
        } else if (pVal < parentVal && qVal < parentVal) {
            // If both p and q are lesser than parent
            return lowestCommonAncestor(root.left, p, q);
        } else {
            // We have found the split point, i.e. the LCA node.
            return root;
        }
    }
}

File: [0235][Easy][Lowest Common Ancestor of a Binary Search Tree]/Lowest_Common_Ancestor_of_a_Binary_Search_Tree.py
#-*- coding: utf-8 -*-  
# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
# solution 1, use recursion to travel all the nodes of the BST.
# leetcode time     cost : 60 ms
# leetcode memory   cost : 20.1 MB
# Time  Complexity: O(N)
# Space Complexity: O(N)
class Solution:
    def lowestCommonAncestor(self, root, p, q):
        """
        :type root: TreeNode
        :type p: TreeNode
        :type q: TreeNode
        :rtype: TreeNode
        """
        # Value of current node or parent node.
        parent_val = root.val

        # Value of p
        p_val = p.val

        # Value of q
        q_val = q.val

        # If both p and q are greater than parent
        if p_val > parent_val and q_val > parent_val:    
            return self.lowestCommonAncestor(root.right, p, q)
        # If both p and q are lesser than parent
        elif p_val < parent_val and q_val < parent_val:    
            return self.lowestCommonAncestor(root.left, p, q)
        # We have found the split point, i.e. the LCA node.
        else:
            return root